usecases/config/runtime/values.go

```diff
@@ -0,0 +1,134 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package runtime
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"gopkg.in/yaml.v3"
+)
+
+// Type represents different types that is supported in runtime configs
+type DynamicType interface {
+	~int | ~float64 | ~bool | time.Duration | ~string
+}
+
+// Value represents any runtime config value. It's zero value is
+// fully usable.
+// If you want zero value with different `default`, use `NewDynamicValue` constructor.
+type DynamicValue[T DynamicType] struct {
+	// val is the dynamically changing value.
+	val atomic.Value
+	// protects `val` during Reset()
+	valMu sync.Mutex
+
+	// def represents the default value.
+	def T
+	// protects `def`
+	defMu sync.Mutex
+}
+
+// NewDynamicValue returns an instance of DynamicValue as passed in type
+// with passed in value as default.
+func NewDynamicValue[T DynamicType](val T) *DynamicValue[T] {
+	return &DynamicValue[T]{
+		def: val,
+	}
+}
+
+// Get returns a current value for the given config. It can either be dynamic value or default
+// value (if unable to get dynamic value)
+// Consumer of the dynamic config value should care only about this `Get()` api.
+func (dv *DynamicValue[T]) Get() T {
+	// Handle zero-value of `*DynamicValue[T]` without panic.
+	if dv == nil {
+		var zero T
+		return zero
+	}
+
+	v := dv.val.Load()
+	if v != nil {
+		return v.(T)
+	}
+
+	dv.defMu.Lock()
+	def := dv.def
+	dv.defMu.Unlock()
+
+	return def
+}
+
+// Reset removes the old dynamic value.
+func (dv *DynamicValue[T]) Reset() {
+	// Once atomic value is set via `Store()`, no api to reset it
+	// hence replacing the full value.
+	dv.valMu.Lock()
+	defer dv.valMu.Unlock()
+
+	dv.val = atomic.Value{}
```

Comment:
why not simply `dv.val.Store(nil)`

Reply:
In fact tried it before. [That would panic. ](https://cs.opensource.google/go/go/+/refs/tags/go1.24.2:src/sync/atomic/value.go;l=47-50)

Reply:
hmm, in that case, we don't need the atomic value at all. You could simplify everything by simply using a RWMock and and a regular value.

Currently, we have duplicate synchronization from atomic + locks

Reply:
True. but lock is just to replace the atomic value. I still like atomic value's nice API to load and store (still better than manual locks, less chances of errors like forgetting to unlock or something).

Currently the valMu is only in one place to replace the value during Reset. I think it's not too bad. I'm happy to leave it this way unless you have different opinions.

Reply:
Ah, there is a race condition now in the `Get()` then because you are reading the `dv.val` without a lock

Reply:
and in `Set()` as well

Reply:
I ended up replacing the atomic.Value -> RWmutex guarded value. Also made the unmarshaling much simpler :)